src/ui/__tests__/ui_app.js

```diff
@@ -0,0 +1,253 @@
+          id: 'uiapp-test-defaults',
+          title: undefined,
+          order: 0,
+          url: '/app/uiapp-test-defaults',
+          subUrlBase: '/app/uiapp-test-defaults',
+          description: undefined,
+          icon: undefined,
+          linkToLastSubUrl: true,
+          hidden: false,
+          disabled: false,
+          tooltip: ''
+        });
+      });
+
+      it('has a built-in default order of 0', () => {
+        expect(newApp.order).to.be(0);
+      });
+
+      it('has a built-in default template name of ui_app', () => {
+        expect(newApp.templateName).to.be('ui_app');
+      });
+    });
+
+    describe('with spec', () => {
+      const spec = getMockSpec();
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('copies the ID from the spec', () => {
+        expect(newApp.id).to.be(spec.id);
+      });
+
+      it('copies field values from spec', () => {
+        // test that the fields exist, but have undefined value
+        expect(newApp.main).to.be(spec.main);
+        expect(newApp.title).to.be(spec.title);
+        expect(newApp.description).to.be(spec.description);
+        expect(newApp.icon).to.be(spec.icon);
+        expect(newApp.linkToLastSubUrl).to.be(spec.linkToLastSubUrl);
+        expect(newApp.templateName).to.be(spec.templateName);
+        expect(newApp.order).to.be(spec.order);
+        expect(newApp.navLink).to.eql({
+          id: 'uiapp-test',
+          title: 'UIApp Test',
+          order: 9000,
+          url: '/app/uiapp-test',
+          subUrlBase: '/app/uiapp-test',
+          description: 'Test of UI App Constructor',
+          icon: 'ui_app_test.svg',
+          linkToLastSubUrl: true,
+          hidden: false,
+          disabled: false,
+          tooltip: ''
+        });
+      });
+    });
+
+    describe('reference fields', () => {
+      const spec = getMockSpec({ testSpec: true });
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('has a reference to the uiExports object', () => {
+        expect(newApp.uiExports).to.be(uiExports);
+      });
+
+      it('has a reference to the original spec', () => {
+        expect(newApp.spec).to.be(spec);
+      });
+
+      it('has a reference to the spec.injectVars function', () => {
+        const helloFunction = () => 'hello';
+        const spec = {
+          id: 'uiapp-test',
+          injectVars: helloFunction
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(helloFunction);
+      });
+    });
+
+    describe('app.getInjectedVars', () => {
+      it('is noop function by default', () => {
+        const spec = {
+          id: 'uiapp-test'
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(noop);
+      });
+    });
+
+    describe('hidden and listed flags', () => {
```

Comment:
When I was reading these tests, I was having a hard time figuring out how these two properties interacted. After I read the source, I saw that listed was dependent on hidden, and that hidden was cast to a boolean values.

I think this structure makes that relationship a little clearer. What do you think?

```javascript
    describe('hidden', () => {
      describe('is cast to boolean value', () => {
        it('when undefined', () => {
          const spec = {
            id: 'uiapp-test',
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.hidden).to.be(false);
        });

        it('when null', () => {
          const spec = {
            id: 'uiapp-test',
            hidden: null,
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.hidden).to.be(false);
        });
      });
    });

    describe('listed', () => {
      describe('defaults to the opposite value of hidden', () => {
        it(`when it's null and hidden is true`, () => {
          const spec = {
            id: 'uiapp-test',
            hidden: true,
            listed: null,
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.listed).to.be(false);
        });

        it(`dhen it's null and hidden is false`, () => {
          const spec = {
            id: 'uiapp-test',
            hidden: false,
            listed: null,
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.listed).to.be(true);
        });

        it(`when it's undefined and hidden is false`, () => {
          const spec = {
            id: 'uiapp-test',
            hidden: false,
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.listed).to.be(true);
        });

        it(`when it's undefined and hidden is true`, () => {
          const spec = {
            id: 'uiapp-test',
            hidden: true,
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.listed).to.be(false);
        });
      });

      it(`is set to true when it's passed as true`, () => {
        const spec = {
          id: 'uiapp-test',
          listed: true,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(true);
      });

      it(`is set to false when it's passed as false`, () => {
        const spec = {
          id: 'uiapp-test',
          listed: false,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(false);
      });
    });
```